buildpackrunner/runner.go

```diff
@@ -193,6 +231,48 @@ func (runner *Runner) downloadBuildpacks() error {
 	return nil
 }
 
+func (runner *Runner) cleanCacheDir() error {
+	if !runner.config.IsMultiBuildpack() {
+		return nil
+	}
+
+	neededCacheDirs := []string{filepath.Join(runner.config.BuildArtifactsCacheDir(), "primary")}
+	buildpacks := runner.config.BuildpackOrder()
+	supplyBuildpacks := buildpacks[0:(len(buildpacks) - 1)]
+
+	for _, bp := range supplyBuildpacks {
+		neededCacheDirs = append(neededCacheDirs, runner.supplyCachePath(bp))
+	}
+
+	dirs, err := ioutil.ReadDir(runner.config.BuildArtifactsCacheDir())
+	if err != nil {
+		return err
+	}
+	var foundCacheDirs []string
+	for _, dirInfo := range dirs {
+		foundCacheDirs = append(foundCacheDirs, filepath.Join(runner.config.BuildArtifactsCacheDir(), dirInfo.Name()))
+	}
+
+	for _, dir := range foundCacheDirs {
+		if runner.dirUnused(dir, neededCacheDirs) {
```

Comment:
This seems like an inefficient way to perform this check as it is O(n^2). I know that this level of performance won't really matter, but can we use something like a Map to perform this check more efficiently?